### ISSUE TITLE
Channel padding in conv2d needs to be zero

### ISSUE THREAD

**Issue by pavlepopovic** [tenstorrent/tt-metal#20768]
### Describe the bug

In a customer model usecase, we have an input image of shape [1, 10, 2048, 128], upon which we need to do a convolution.

First thing we need to do is pad input channels up to 16 (in order to perform conv), and it seems that the padding must be a zero.
If it needs to be zero -> we must insert explicit pad op which will fill the values with zero (and copy the tensor)
If it's not zero -> it appears not to be working. Is this a limitation or a bug?
Attaching a unit test here:
The first conv will run with zero-padded channels, resulting in a PCC pass.
The second conv will run with padding of 1520 (picked it randomly), resulting in a PCC failure.
For simplicity, the second conv will also call pad op (even tho in the model we will do something else, and padding will be random, constant value makes debugging easier)

There are also commented out portions in the code, which will print the conv input tensor:
- with all 16 channels (10 actual values, and 6 padded ones)
- with 10 actual channels (6 padded ones are not printed)

Comparing input tensors from first and second conv runs reveals that the difference is only in the padded channels.

```
def print_tensor_to_file(tensor: torch.Tensor, file_path: str, lines_per_block: int = 4096, remove_last_n: int = 0):
    """
    Prints a torch tensor to a file with formatting:
    - Each row is from the last dimension (flattened across all other dimensions).
    - Removes the last `remove_last_n` elements from each row before printing.
    - Adds a newline after each row.
    - After every `lines_per_block` rows, inserts 5 blank lines.

    Args:
        tensor (torch.Tensor): The input tensor.
        file_path (str): Path to the output file.
        lines_per_block (int): Number of lines after which to insert 5 newlines.
        remove_last_n (int): Number of elements to remove from the end of each row.
    """
    with open(file_path, "w") as f:
        reshaped = tensor.reshape(-1, tensor.shape[-1])
        end_idx = None if remove_last_n == 0 else -remove_last_n

        for i, row in enumerate(reshaped):
            trimmed = row[:end_idx] if end_idx is not None else row
            print(" ".join(map(str, trimmed.tolist())), file=f)
            if (i + 1) % lines_per_block == 0:
                print("\n" * 5, file=f)


@pytest.mark.parametrize("device_params", [{"l1_small_size": 2 * 16384}], indirect=True)
@pytest.mark.parametrize(
    "output_channels, input_channels, input_height, input_width",
    ((64, 10, 2048, 128),),
)
@pytest.mark.parametrize(
    "weights_dtype",
    [ttnn.bfloat8_b],
)
@pytest.mark.parametrize(
    "activations_dtype",
    [ttnn.bfloat8_b],
)
@pytest.mark.parametrize(
    "filter, pad, stride",
    [
        [4, 1, 2],
    ],
)
@pytest.mark.parametrize(
    "shard_height",
    [
        (512),
    ],
)
def test_run_conv_with_padding(
    device,
    output_channels,
    input_channels,
    input_height,
    input_width,
    filter,
    pad,
    stride,
    weights_dtype,
    activations_dtype,
    shard_height,
):
    if device.core_grid.y != 8 and is_wormhole_b0():
        pytest.skip("Needs 8x8 grid for wormhole_b0")

    torch.manual_seed(2143)

    stride_h = stride
    stride_w = stride
    pad_h = pad
    pad_w = pad
    filter_h = filter
    filter_w = filter
    fp32_accum = False
    packer_l1_acc = False
    groups = 1

    conv_input_shape = (1, input_channels, input_height, input_width)
    conv_weight_shape = (output_channels, input_channels // groups, filter_h, filter_w)
    conv_bias_shape = (1, 1, 1, output_channels)

    torch_input_tensor_nchw = torch.randn(conv_input_shape, dtype=torch.bfloat16).float()
    torch_weight_tensor = torch.randn(conv_weight_shape, dtype=torch.bfloat16).float()
    torch_bias_tensor = torch.randn(conv_bias_shape, dtype=torch.bfloat16).float()
    tt_bias_tensor = ttnn.from_torch(
        torch_bias_tensor, weights_dtype if weights_dtype != ttnn.bfloat8_b else ttnn.float32
    )
    torch_bias_tensor = torch_bias_tensor.reshape(-1)
    torch_out_golden_tensor = torch.nn.functional.conv2d(
        torch_input_tensor_nchw,
        torch_weight_tensor,
        bias=torch_bias_tensor,
        stride=(stride_h, stride_w),
        padding=(pad_h, pad_w),
        groups=groups,
    )
    torch_out_golden_tensor = torch.nn.functional.relu(torch_out_golden_tensor)

    tt_weight_tensor = ttnn.from_torch(
        torch_weight_tensor, weights_dtype if weights_dtype != ttnn.bfloat8_b else ttnn.float32
    )

    tt_input_tensor_on_host = ttnn.from_torch(torch_input_tensor_nchw, dtype=ttnn.bfloat16)
    num_cores = 64

    shard_shape = [shard_height, input_width]
    core_grid = device.compute_with_storage_grid_size()

    shard_grid = ttnn.num_cores_to_corerangeset(num_cores, core_grid)
    shard_spec = ttnn.ShardSpec(
        shard_grid,
        shard_shape,
        ttnn.ShardOrientation.ROW_MAJOR,
    )

    l1_sharded_config = ttnn.MemoryConfig(ttnn.TensorMemoryLayout.HEIGHT_SHARDED, ttnn.BufferType.L1, shard_spec)

    tt_input_tensor_on_device = ttnn.to_device(tt_input_tensor_on_host, device=device, memory_config=l1_sharded_config)

    # pad up to 16 channels with zeros
    nchw = ttnn.pad(
        tt_input_tensor_on_device, ((0, 0), (0, 6), (0, 0), (0, 0)), value=0.0, memory_config=l1_sharded_config
    )
    nhwc = ttnn.permute(nchw, (0, 2, 3, 1))
    nhwc = ttnn.reshape(nhwc, [1, 1, nhwc.shape[0] * nhwc.shape[1] * nhwc.shape[2], nhwc.shape[-1]])

    conv_config = ttnn.Conv2dConfig(
        dtype=activations_dtype,
        weights_dtype=weights_dtype,
        shard_layout=ttnn.TensorMemoryLayout.HEIGHT_SHARDED,
        input_channels_alignment=16,
        deallocate_activation=False,
        enable_split_reader=True,
        enable_subblock_padding=False,
        output_layout=ttnn.TILE_LAYOUT,
        activation="relu",
        reshard_if_not_optimal=False,
        act_block_h_override=32 * 16,
        enable_act_double_buffer=True,
    )
    compute_config = ttnn.init_device_compute_kernel_config(
        device.arch(),
        math_fidelity=ttnn.MathFidelity.HiFi4,
        fp32_dest_acc_en=fp32_accum,
        packer_l1_acc=packer_l1_acc,
    )

    #nhwc_on_host = ttnn.to_torch(nhwc, mesh_composer=None)
    #print_tensor_to_file(nhwc_on_host, "tensor_zero_padded.txt", input_height * input_width // num_cores, 0)
    #print_tensor_to_file(nhwc_on_host, "tensor_zero_padded_padding_removed.txt", input_height * input_width // num_cores, 6) # remove padded channels
    [out, [out_h, out_w]] = ttnn.conv2d(
        input_tensor=nhwc,
        weight_tensor=tt_weight_tensor,
        bias_tensor=tt_bias_tensor,
        device=device,
        in_channels=input_channels,
        out_channels=output_channels,
        input_height=input_height,
        input_width=input_width,
        batch_size=1,
        kernel_size=[filter_h, filter_w],
        stride=[stride_h, stride_w],
        padding=[pad_h, pad_w],
        conv_config=conv_config,
        compute_config=compute_config,
        groups=1,
        return_output_dim=True,
        return_weights_and_bias=False,
    )

    tt_output_tensor = ttnn.from_device(out)
    torch_output_tensor = ttnn.to_torch(tt_output_tensor, mesh_composer=None)
    torch_output_tensor = torch_output_tensor.reshape(1, out_h, out_w, torch_output_tensor.shape[-1])
    torch_output_tensor = torch_output_tensor[:, :, :, :output_channels]
    torch_output_tensor = torch.permute(torch_output_tensor, (0, 3, 1, 2))

    target_pcc = 0.99
    print("Starting pcc check with zero padded channels...")
    passing, pcc_msg = check_with_pcc_without_tensor_printout(
        torch_output_tensor, torch_out_golden_tensor, pcc=target_pcc
    )
    print("PCC is: ", pcc_msg)
    assert passing

    # pad to 16 channels (avoid padding with zeros, use random padded numbers)
    nchw_padded_with_non_zero = ttnn.pad(
        tt_input_tensor_on_device, ((0, 0), (0, 6), (0, 0), (0, 0)), value=1520.0, memory_config=l1_sharded_config
    )
    nhwc_reshaped = ttnn.permute(nchw_padded_with_non_zero, (0, 2, 3, 1))
    nhwc_reshaped = ttnn.reshape(
        nhwc_reshaped,
        [1, 1, nhwc_reshaped.shape[0] * nhwc_reshaped.shape[1] * nhwc_reshaped.shape[2], nhwc_reshaped.shape[-1]],
    )

    #nhwc_reshaped_on_host = ttnn.to_torch(nhwc_reshaped, mesh_composer=None)
    #print_tensor_to_file(nhwc_reshaped_on_host, "tensor_non_zero_padded.txt", input_height * input_width // num_cores, 0)
    #print_tensor_to_file(nhwc_reshaped_on_host, "tensor_non_zero_padded_padding_removed.txt", input_height * input_width // num_cores, 6)
    [out, [out_h, out_w]] = ttnn.conv2d(
        input_tensor=nhwc_reshaped,
        weight_tensor=tt_weight_tensor,
        bias_tensor=tt_bias_tensor,
        device=device,
        in_channels=input_channels,
        out_channels=output_channels,
        input_height=input_height,
        input_width=input_width,
        batch_size=1,
        kernel_size=[filter_h, filter_w],
        stride=[stride_h, stride_w],
        padding=[pad_h, pad_w],
        conv_config=conv_config,
        compute_config=compute_config,
        groups=1,
        return_output_dim=True,
        return_weights_and_bias=False,
    )

    tt_output_tensor = ttnn.from_device(out)
    torch_output_tensor = ttnn.to_torch(tt_output_tensor, mesh_composer=None)
    torch_output_tensor = torch_output_tensor.reshape(1, out_h, out_w, torch_output_tensor.shape[-1])
    torch_output_tensor = torch_output_tensor[:, :, :, :output_channels]
    torch_output_tensor = torch.permute(torch_output_tensor, (0, 3, 1, 2))

    target_pcc = 0.99
    print("Starting PCC check with non-zero padded channels...")
    passing, pcc_msg = check_with_pcc_without_tensor_printout(
        torch_output_tensor, torch_out_golden_tensor, pcc=target_pcc
    )
    print("PCC is: ", pcc_msg)
    assert passing
```

**Comment by pavlepopovic** [comment]
Switching activations_dtype to float16 makes the test pass, might be related to this:
https://github.com/tenstorrent/tt-metal/issues/20808

**Comment by pavlejosipovic** [comment]
We didn't come up with a performant solution for this one.
I think we can get best perf if you pad up the input tensor while it's still in N,C,HW format (instead of NHW, C), or just use float16 as output dtype on this conv?

**Comment by pavlepopovic** [comment]
Can use float16, the following conv is read-stick bound, so not much gained from bfp8 in the first place

**Comment by pavlepopovic** [comment]
yes
